core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala

```diff
@@ -0,0 +1,413 @@
+          session.selectEventsForBucket(missing.bucket, missing.previousOffset, Some(missing.maxOffset))
+            .onComplete(newResultSetCb.invoke)
+        }
+      }
+
+      @tailrec def tryPushOne(): Unit = {
+        log.debug("TryPushOne() {} {}", queryState, missingLookup)
+        queryState match {
+          case QueryResult(rs) if isAvailable(out) =>
+            if (rs.isExhausted) {
+              queryExhausted()
+            } else if (rs.getAvailableWithoutFetching == 0) {
+              log.debug("Fetching more")
+              fetchMore(rs)
+            } else if (missingLookup.isDefined) {
+              val m = missingLookup.get
+              val event = extractEvent(rs.one())
+              if (event.persistentRepr.persistenceId == m.persistenceId && m.missing.contains(event.tagPidSequenceNr)) {
+                val remainingEvents = m.missing - event.tagPidSequenceNr
+                log.info("{}: Found a missing event, sequence nr {}. Remaining missing: {}", session.tag, event.tagPidSequenceNr, remainingEvents)
+                if (remainingEvents.isEmpty) {
+                  stopLookingForMissing(m, event :: m.buffered)
+                } else {
+                  log.info("{}: There are more missing events. {}", session.tag, remainingEvents)
+                  missingLookup = missingLookup.map(m => m.copy(
+                    missing = remainingEvents,
+                    buffered = event :: m.buffered
+                  ))
+                }
+              }
+              tryPushOne()
+            } else {
+              rowsForCurrentQuery += 1
+              val row = rs.one()
+              val repr = extractEvent(row)
+              val pid = repr.persistentRepr.persistenceId
+              val (expectedSequenceNr, shouldFail, previousUUID) = tagPidSequenceNrs.get(pid) match {
+                case None =>
+                  val currentBucketStart = UUIDs.startOf(currTimeBucket.startTimestamp)
+                  val startingOffset: UUID = if (UUIDComparator.comparator.compare(currentBucketStart, fromOffset) < 0) {
+                    log.debug("Starting at fromOffset")
+                    fromOffset
+                  } else {
+                    log.debug("Starting at startOfBucket")
+                    currentBucketStart
+                  }
+                  (1L, false, startingOffset)
+                case Some((lastSequenceNr, lastUUID)) =>
+                  (lastSequenceNr + 1, true, lastUUID)
+              }
+
+              if (repr.tagPidSequenceNr < expectedSequenceNr) {
+                log.warning("Got duplicate sequence number. Persistence id: {}. Expected sequence nr: {}. Actual {}. This will be dropped.", pid, expectedSequenceNr, repr.tagPidSequenceNr)
+                tryPushOne()
+              } else if (repr.tagPidSequenceNr > expectedSequenceNr) {
```

Comment:
let's see, here is when we switch to the "look for missing" mode.
how far back do we look when it's a new persistenceId? only from `previousUUID`?
I'm thinking if it's easy to miss event 1, 2, when receiving 3 as the first?

Reply:
`previousUUID` is actually the previous uuid for that tag/pid combination so we look back as far as we last saw an event (tho only this bucket and the last).

For a new persistenceId we don't have one so the `previousUUID` is set to the start of the current bucket (or the offset if that is later). I actually think we should make this the start of the previousBucket to make it inline with the case where it isn't the first persistenceId, WDYT?

Reply:
that is good, I agree that it should scan entire previousBucket for new persistenceId

Reply:
k, will do this as part of this PR, it was an oversight